### PR TITLE
Proposal: Dual-Key Stealth Address Protocol

### PR DESCRIPTION
**Pull request by wuyahuang**
### Project Abstract

Public ledgers are generally seen as “pseudo-anonymous” as addresses can be linked to one person although that person’s identity is unknown to the public. However, by combining this info with other data it is possible to discover the real world identity behind the address. Many individuals and companies prefer to have an additional layer of security in order to keep their privacy. That’s where DKSAP come to play.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying?

- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] I have read and understood the [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md), [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/grant_guidelines_per_category.md) and [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md).
- [x] A BTC, Ethereum (USDT/USDC/DAI) or Polkadot/Kusama (aUSD) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledge the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### How Did You Hear About our grants program?

- [ ] Social Media
- [ ] Hackathon
- [ ] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [x] Online Search
- [ ] Other: _______

### CODE REVIEW

File: applications/DKSAP.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# W3F Grant Proposal
+
+- **Project Name:** DKSAP
+- **Team Name:** DKSAP
+- **Payment Address:** BTC, Ethereum (USDT/USDC/DAI) or Polkadot/Kusama (aUSD) payment address. Please also specify the currency. (e.g. 0x8920... (DAI))
```
Comment (laboon) [inline applications/DKSAP.md:5]: This should be filled in as part of the original grant proposal PR.

File: applications/DKSAP.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+### Milestone 1 — Node.js features
+
+- **Estimated duration:** 4 weeks
+- **FTE:**  1
+- **Costs:** 5,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | Apache 2.0 / GPLv3 / MIT / Unlicense |
+| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
```
Comment (laboon) [inline applications/DKSAP.md:96]: I'm not sure what exactly you are proposing to deliver here. You mention Solidity in several places, but also Substrate nodes - is this meant to be a pallet? Solidity code? Solidity code which requires specific precompiles in Substrate? Note that we generally do not accept Solidity code as we are not experts in reviewing it. We would be interested in an ink! version, however.

The milestone deliverables for Milestone 1 and 2 should also be more specific - what specific features will be included and how will they be accessed? These will be used when we are verifying the deliverables, so more detail is generally better than less. If you look through some of the accepted grant PRs, you can see some examples of what is considered sufficient detail.

**Review by laboon — CHANGES_REQUESTED** [review]
I think DKSAP support is potentially interesting, but would like more detail on what exactly you are proposing (see comments) before approving.

**Comment by wuyahuang** [comment]
Hi @Noc2 , Thanks for you feedback.

**Review by Noc2 — DISMISSED** [review]
Thanks for the application. I’m happy to support the application. I have just one question: where can I find the solidity implementation?

**Comment by wuyahuang** [comment]
Hi, @Noc2 . Thank you for your affirmation and encouragement. The DKSAP has been implemented in a number of cryptocurrency systems, including Monero, Samourai Wallet, and TokenPay. I'm afraid there is not a solidity implementation of DKSAP yet, but if polkadot ecosystems likes it, I'm happy to develop implementation for solidity. 😄

**Comment by Noc2** [comment]
Ah, I was referring to this: “After the Solidity implementation of DKSAP is fully tested”... but you probably meant “after the ink! Implementation”. In any case, we prefer the ink! Implementation, but you potentially can apply for solidity implementation here: https://moonbeam.foundation/grants/

**Comment by wuyahuang** [comment]
Oops. That's a typo. I have already updated it. I'm happy to apply for solidity implemention after this GRANT delivered.

**Review by Noc2 — APPROVED** [review]

**Comment by laboon** [comment]
Thanks for the updates @wuyahuang , I am happy to approve now.

**Review by laboon — APPROVED** [review]

**Comment by wuyahuang** [comment]
Thanks for your approvals @Noc2 @laboon , I'm looking forward to delivering DKSAP implementation to the Polkadot community in 6 weeks, and can't wait to see positive results.

**Pull request merged**